File: win-prog/ProgWinCS/Printing/SimplePrintDialog/SimplePrintDialog.cs
//------------------------------------------------
// SimplePrintDialog.cs © 2001 by Charles Petzold
//------------------------------------------------
using System;
using System.Drawing;
using System.Drawing.Printing;
using System.Windows.Forms;

class PrintDialogHelloWorld: Form
{
     const int iNumberPages = 3;
     int       iPagesToPrint, iPageNumber;

     public static void Main()
     {
          Application.Run(new PrintDialogHelloWorld());
     }
     public PrintDialogHelloWorld()
     {
          Text = "Simple PrintDialog";

          Menu = new MainMenu();
          Menu.MenuItems.Add("&File");
          Menu.MenuItems[0].MenuItems.Add("&Print...", 
                                   new EventHandler(MenuFilePrintOnClick));
     }
     void MenuFilePrintOnClick(object obj, EventArgs ea)
     {
               // Create the PrintDocument and PrintDialog.

          PrintDocument prndoc = new PrintDocument();
          PrintDialog   prndlg = new PrintDialog();
          prndlg.Document = prndoc;

               // Allow a page range.

          prndlg.AllowSomePages = true;
          prndlg.PrinterSettings.MinimumPage = 1;
          prndlg.PrinterSettings.MaximumPage = iNumberPages;
          prndlg.PrinterSettings.FromPage = 1;
          prndlg.PrinterSettings.ToPage = iNumberPages;

               // If the dialog box returns OK, print.

          if(prndlg.ShowDialog() == DialogResult.OK)
          {
               prndoc.DocumentName = Text;
               prndoc.PrintPage += new PrintPageEventHandler(OnPrintPage);

                    // Determine which pages to print.

               switch (prndlg.PrinterSettings.PrintRange)
               {
               case PrintRange.AllPages:
                    iPagesToPrint = iNumberPages;
                    iPageNumber = 1;
                    break;

               case PrintRange.SomePages:
                    iPagesToPrint = 1 + prndlg.PrinterSettings.ToPage -
                                        prndlg.PrinterSettings.FromPage;
                    iPageNumber = prndlg.PrinterSettings.FromPage;
                    break;
               }
               prndoc.Print();
          }
     }
     void OnPrintPage(object obj, PrintPageEventArgs ppea)
     {
          Graphics grfx  = ppea.Graphics;
          Font     font  = new Font("Times New Roman", 360);
          string   str   = iPageNumber.ToString();
          SizeF    sizef = grfx.MeasureString(str, font);

          grfx.DrawString(str, font, Brushes.Black, 
                    (grfx.VisibleClipBounds.Width - sizef.Width) / 2,
                    (grfx.VisibleClipBounds.Height - sizef.Height) / 2);

          iPageNumber += 1;
          iPagesToPrint -= 1;
          ppea.HasMorePages = iPagesToPrint > 0;
     }
}

File: win-prog/ProgWinVB/Printing/SimplePrintDialog/SimplePrintDialog.vb
'--------------------------------------------------
' SimplePrintDialog.vb (c) 2002 by Charles Petzold
'--------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Drawing.Printing
Imports System.Windows.Forms

Class SimplePrintDialog
    Inherits Form

    Const iNumberPages As Integer = 3
    Private iPagesToPrint, iPageNumber As Integer

    Shared Sub Main()
        Application.Run(New SimplePrintDialog())
    End Sub

    Sub New()
        Text = "Simple PrintDialog"

        Menu = New MainMenu()
        Menu.MenuItems.Add("&File")
        Menu.MenuItems(0).MenuItems.Add("&Print...", _
                                        AddressOf MenuFilePrintOnClick)
    End Sub

    Private Sub MenuFilePrintOnClick(ByVal obj As Object, _
                                     ByVal ea As EventArgs)
        ' Create the PrintDocument and PrintDialog.
        Dim prndoc As New PrintDocument()
        Dim prndlg As New PrintDialog()
        prndlg.Document = prndoc

        ' Allow a page range.
        prndlg.AllowSomePages = True
        prndlg.PrinterSettings.MinimumPage = 1
        prndlg.PrinterSettings.MaximumPage = iNumberPages
        prndlg.PrinterSettings.FromPage = 1
        prndlg.PrinterSettings.ToPage = iNumberPages

        ' If the dialog box returns OK, print.
        If prndlg.ShowDialog() = DialogResult.OK Then
            prndoc.DocumentName = Text
            AddHandler prndoc.PrintPage, AddressOf OnPrintPage

            ' Determine which pages to print.
            Select Case prndlg.PrinterSettings.PrintRange
                Case PrintRange.AllPages
                    iPagesToPrint = iNumberPages
                    iPageNumber = 1

                Case PrintRange.SomePages
                    iPagesToPrint = 1 + prndlg.PrinterSettings.ToPage - _
                                    prndlg.PrinterSettings.FromPage()
                    iPageNumber = prndlg.PrinterSettings.FromPage
            End Select
            prndoc.Print()
        End If
    End Sub

    Private Sub OnPrintPage(ByVal obj As Object, _
                            ByVal ppea As PrintPageEventArgs)
        Dim grfx As Graphics = ppea.Graphics
        Dim fnt As New Font("Times New Roman", 360)
        Dim str As String = iPageNumber.ToString()
        Dim szf As SizeF = grfx.MeasureString(str, fnt)

        grfx.DrawString(str, fnt, Brushes.Black, _
                        (grfx.VisibleClipBounds.Width - szf.Width) / 2, _
                        (grfx.VisibleClipBounds.Height - szf.Height) / 2)
        iPageNumber += 1
        iPagesToPrint -= 1
        ppea.HasMorePages = iPagesToPrint > 0
    End Sub
End Class
